core/src/main/java/org/elasticsearch/gateway/PrimaryShardAllocator.java

```diff
@@ -95,11 +99,7 @@ public boolean allocateUnassigned(RoutingAllocation allocation) {
                 continue;
             }
 
-            final IndexMetaData indexMetaData = metaData.getIndexSafe(shard.index());
-            // don't go wild here and create a new IndexSetting object for every shard this could cause a lot of garbage
-            // on cluster restart if we allocate a boat load of shards
-            if (shard.allocatedPostIndexCreate(indexMetaData) == false) {
-                // when we create a fresh index
+            if (shard.recoverySource() != EXISTING_STORE && shard.recoverySource() != SNAPSHOT) {
```

Comment:
LOCAL_SHARDS is treated from the point of view of the primary shard allocator in the same way as new index creation. It does not need to look for existing shard data.